### PR TITLE
Proofs for keyring_trace module eq, clear and clean_up functions

### PR DESCRIPTION
**Pull request by tegansiobhan**
Proofs and harnesses for aws_cryptosdk_keyring_trace_eq, aws_cryptosdk_keyring_trace_clear and aws_cryptosdk_keyring_trace_clean_up functions. 

Added corresponding assumptions to the source code. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tegansiobhan** [comment]
@dsn @dougch

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clean_up/Makefile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+include ../Makefile.string 
+include ../Makefile.aws_array_list
+
+CBMC_UNWINDSET += --unwindset aws_cryptosdk_keyring_trace_is_valid.0:$(shell echo $$((1 + $(MAX_ITEM_SIZE)))),ensure_trace_has_allocated_records.0:$(shell echo $$((1 + $(MAX_ITEM_SIZE))))
+
+CBMCFLAGS +=
+
+ENTRY = aws_cryptosdk_keyring_trace_clean_up_harness
+
+DEPENDENCIES +=	$(SRCDIR)/c-enc-sdk-src/keyring_trace.c
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clean_up/Makefile]: Sort these

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clean_up/aws_cryptosdk_keyring_trace_clean_up_harness.c [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * permissions and limitations under the License.
+ */
+
+#include <aws/cryptosdk/private/keyring_trace.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void aws_cryptosdk_keyring_trace_clean_up_harness() {
+    /* data structure */
+    struct aws_array_list trace;
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clean_up/aws_cryptosdk_keyring_trace_clean_up_harness.c]: assuming non-null?

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clear/Makefile [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+###########
+# if Makefile.local exists, use it. This provides a way to override the defaults 
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+include ../Makefile.string
+include ../Makefile.aws_array_list
+
+CBMC_UNWINDSET +=  --unwindset aws_cryptosdk_keyring_trace_is_valid.0:$(shell echo $$((1 + $(MAX_ITEM_SIZE)))),ensure_trace_has_allocated_records.0:$(shell echo $$((1 + $(MAX_ITEM_SIZE))))
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clear/Makefile]: These can each go on their own line.

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clear/Makefile [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+CBMCFLAGS +=
+
+ENTRY = aws_cryptosdk_keyring_trace_clear_harness
+
+DEPENDENCIES +=	$(SRCDIR)/c-enc-sdk-src/keyring_trace.c
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_clear/Makefile]: Sort this

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_eq/aws_cryptosdk_keyring_trace_eq_harness.c [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ * permissions and limitations under the License.
+ */
+
+#include <aws/cryptosdk/private/keyring_trace.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void aws_cryptosdk_keyring_trace_eq_harness() {
+    /* data structure */
+    struct aws_array_list lhs;
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_eq/aws_cryptosdk_keyring_trace_eq_harness.c]: assuming non-null

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_eq/aws_cryptosdk_keyring_trace_eq_harness.c [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    __CPROVER_assume(aws_array_list_is_valid(&lhs));
+    ensure_trace_has_allocated_records(&lhs, MAX_STRING_LEN);
+    __CPROVER_assume(aws_cryptosdk_keyring_trace_is_valid(&lhs));
+
+    __CPROVER_assume(aws_array_list_is_bounded(&rhs, MAX_INITIAL_ITEM_ALLOCATION, MAX_ITEM_SIZE));
+    __CPROVER_assume(rhs.item_size == sizeof(struct aws_cryptosdk_keyring_trace_record));
+    ensure_array_list_has_allocated_data_member(&rhs);
+    __CPROVER_assume(aws_array_list_is_valid(&rhs));
+    ensure_trace_has_allocated_records(&rhs, MAX_STRING_LEN);
+    __CPROVER_assume(aws_cryptosdk_keyring_trace_is_valid(&rhs));
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_eq/aws_cryptosdk_keyring_trace_eq_harness.c]: Using the type-safe `array_list` stubs could make this proof much faster.

Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_eq/aws_cryptosdk_keyring_trace_eq_harness.c]: But if its fast enough, not worth reworking it.

File: source/keyring_trace.c [hunk 7]
```diff
@@ -214,6 +214,10 @@ static bool aws_cryptosdk_keyring_trace_record_eq(
 }
 
 bool aws_cryptosdk_keyring_trace_eq(const struct aws_array_list *a, const struct aws_array_list *b) {
+    AWS_FATAL_PRECONDITION(aws_cryptosdk_keyring_trace_is_valid(a));
+    AWS_FATAL_PRECONDITION(aws_cryptosdk_keyring_trace_is_valid(b));
+    AWS_FATAL_PRECONDITION(a->item_size == sizeof(struct aws_cryptosdk_keyring_trace_record));
```
Comment (danielsn) [inline source/keyring_trace.c]: This seems like something that could be part of the `is_valid()` function

**Comment by tawdry-audrey** [comment]
Rebased, updated, and merged in https://github.com/aws/aws-encryption-sdk-c/pull/494.

**Pull request closed without merging**